python/ray/worker.py

```diff
@@ -404,6 +404,16 @@ def retrieve_and_deserialize(self, object_ids, timeout, error_timeout=10):
                                 object_ids[i + j],
                                 serialization_context,
                             ))
+                    with self.plasma_client.lock:
+                        # The `release` method of C++ plasma client is used in
+                        # `PlasmaBuffer.__dealloc__`, thus this needs to be
+                        # protected by a lock.
+                        # TODO(hchen): This doesn't work in Python 2. Because
+                        # in Python 2, the `__dealloc__` method doesn't get
+                        # called right after the reference count decreases to
+                        # 0. We need to fix this by making the underlying
+                        # Cython or C++ plasma client thread-safe.
+                        del metadata_data_pairs
```

Comment:
This just decrements a reference count, right? But the deallocation won't actually happen here because `metadata` and `data` are still in scope.

If the retrieved value is a numpy array, then the `__dealloc__` won't happen until the object ID goes out of scope (once the user is done using it).

I don't see how to protect the release call with a lock other than by putting a lock in the actual `__dealloc__` dealloc method.

Reply:
`metadata` and `data` should be out of scope, because the `del` is out of the for loop.
However, it the deserialized numpy array still holds the PlasmaBuffer, that's indeed still a problem.

Also I'm still confused why Python 3 could work before this PR. `__dealloc__` should be called without lock.  Do you have any idea?

Maybe I can just skip the test for Python 2 in this PR?

Reply:
@raulchen, it's failing in both Python 2 and Python 3, isn't it? Is it really just Python 2?

We can skip the test in Python 2 if you think that's best, but please add a TODO and note that this is a bug and that it should be re-enabled when we fix it in Plasma.

Reply:
Yeah, it only fails in Python 2.  I confirmed this again by running it 50 times on my machine.